lib/dnsimple/client.rb

```diff
@@ -168,26 +168,41 @@ def request(method, path, data = nil, options = {})
 
     private
 
+    def request_options(custom_options = {})
+      options = base_options
+      Extra.deep_merge!(options, auth_options)
+      Extra.deep_merge!(options, proxy_options)
+      Extra.deep_merge!(options, custom_options)
+    end
+
     def base_options
-      options = {
+      {
           format:   :json,
           headers:  {
               'Accept' => 'application/json',
               'User-Agent' => user_agent,
           },
       }
+    end
+
+    def proxy_options
+      options = {}
 
       if proxy
         address, port = proxy.split(":")
-        options.merge!(http_proxyaddr: address, http_proxyport: port)
+        options = { http_proxyaddr: address, http_proxyport: port }
       end
 
+      options
```

Comment:
To avoid allocating an Hash and then replacing it immediately, I suggest to use an if/else

```
def proxy_options
  if proxy
    address, port = proxy.split(":")
    options = { http_proxyaddr: address, http_proxyport: port }
  else
    {}
  end
end
```

Even better, we can modify the `_options` methods to accept the current options hash, and modify it in place. In this way, we don't allocate unnecessary hashes (such as an empty hash) when not required.

This changeset adds a reasonable allocation overhead compared to the code that was in place before.